torch/distributed/pipelining/schedules.py

```diff
@@ -1041,6 +1041,77 @@ def __init__(
                 "Simply stop passing it, and everything should still work fine."
             )
 
+    def seeded_module_init(
+        self,
+        stage_initializer: Callable[[torch.nn.Module], None],
+        mode: str = "serial",
+        initial_seed: int = 0,
+    ):
+        """
+        Call 'stage_initializer(stage.submod) once per stage owned by this schedule, but configure the RNG seeds before
+        doing so.
+
+        If mode is "parallel", RNG seed is configured once per Pipeline RANK to a value of 'initial_seed' + 'PP RANK'.
+
+        If mode is 'serial', stage 0 uses seed 'initial_seed', then the RNG state is read out after each stage init and
+        transferred to the rank holding the next stage and set before performing next stage init.  Note: when two stages
+        are colocated on the same rank (See Note: V-schedule) there is no need to transfer the RNG state.
+
+        Currently, supports CPU and CUDA RNG states.  Uses device index specified by the 'stage'.  In the future, we may
+        change how 'device' is passed in.
+        """
+        assert mode in ("parallel", "serial"), f"unsupported mode {mode}"
+        if mode == "parallel":
+            seed = initial_seed + self.rank
+            torch.manual_seed(seed)
+            logger.info(
+                "Pipeline using parallel init: rank %d using seed=%d", self.rank, seed
+            )
+
+        local_stage_indices = set({stage.stage_index for stage in self._stages})
+
+        for stage in self._stages:
+            next_stage_is_rank_local = stage.stage_index + 1 in local_stage_indices
+            prev_stage_is_rank_local = stage.stage_index - 1 in local_stage_indices
+            device = stage.device
+
+            # recv, set RNG seed
+            if mode == "serial":
+                if stage.is_first:
+                    torch.manual_seed(initial_seed)
+                    logger.info(
+                        "Pipeline using serial init: rank %d stage %d using manual seed %d and sending RNG state to next rank...",
+                        self.rank,
+                        stage.stage_index,
+                        initial_seed,
+                    )
+                elif not prev_stage_is_rank_local:
+                    # need to know the size/dtype of the state tensors so might as well get current state then overwrite
+                    state_cpu = torch.get_rng_state().to(device)
+                    state_cuda = torch.cuda.get_rng_state(device=device).to(device)
+                    torch.distributed.recv(
```

Comment:
is it overkill to mess with the CPU rng state?  I included it since in theory, someone could use CPU side to calculate something and then copy that to GPU for their tensor init.